### ISSUE TITLE
Ability to define execution sequence

### ISSUE THREAD

**Issue by paulromano** [mit-crpg/openmc#789]
I'd like to propose a new way to define what OpenMC does when it runs. Right now, this is determined in a pretty unsystematic manner:
- To generate plots, one has to run with a `--plot` command line flag.
- To run an eigenvalue of fixed-source calculation, the `Settings.run_mode` attribute is set accordingly -- note that there are two separate XML elements under the hood, `<eigenvalue>` and `<fixed_source>`, so the mapping from Python to XML is weird in this instance. 
- To run a stochastic volume calculation, the `Settings.volume_calculation` attribute has to be set. However, it always runs as part of an eigenvalue or fixed source calculation.
- To run a particle restart, the `--restart` command line flag is used.

Instead of this mix of methods, I propose adding a new `Settings.execute` attribute that defines a list of actions to execute, e.g.
```Python
settings.execute = ['plot', 'volume', 'eigenvalue']
```
would generate plots, run a volume calculation, and then run a k-eigenvalue calculation. The corresponding XML would look like:
```XML
<settings>
  <execute>plot volume eigenvalue</execute>
  ...
</settings>
```
At the same time, I would suggest making `<particles>`, `<batches>`, and `<inactive>` children of `<settings>` rather than `<eigenvalue>` or `<fixed_source>`. I think this scheme is more methodical and extensible than what is currently in place. It also brings the Python and XML representation closer to one another. Thoughts?

**Comment by wbinventor** [comment]
I like your final idea of making `<particles>`, `<batches>` and `<inactive>` children of `<settings>`.

I'm not particularly convinced of the value for an "action item list" if it only includes very high level actions like plotting, volume calcs, and eigenvalue/fixed source simulations. However, in my experience I very rarely want to plot the geometry AND run an eigenvalue calculation at the same time. I've not yet used the volume calculator, but my guess is that this observation would remain true for many if not *most* use cases. I typically see users run multiple simulations with the same geometry, but with different runtime parameters (*e.g.*, number of histories), materials (*e.g.*, fuel enrichments, coolant densities) or temperatures. In these cases, one would not want to clutter up the simulation workflow with plotting and/or volume calculations in between each and every eigenvalue / fixed source simulation - it's more efficient to specify on the command line the run mode, rather than to toggle this in Python / XML all the time.

To be clear, I'm not opposed to making this change, but don't know that it serves any real value as currently proposed. I would be a stronger proponent of this concept if it encompassed lower level functions like "input geometry," "set material", "simulate particle" to make a static library useful.

**Comment by wbinventor** [comment]
Also it is worthwhile to note that the execution sequence concept can already be realized in Python with openmc.run(...), openmc.plot(...), etc.

**Comment by paulromano** [comment]
All good points @wbinventor. Volume calculations are the awkward one right now -- I do want a way to run just a volume calculation (I guess it's my fault for not having that in the first place). Longer term I would like to have finer control of execution, like you suggest, being able to change materials / temperatures, etc. My current thought for how to do that is:
- Have OpenMC define a C API. @RonRahaman and I have discussed a few functions that will allow an external code to get/set temperatures and densities. There are other possibilities too (e.g., changing nuclide densities for depletion).
- Build some kind of Python bindings on top of the C API that will allow direct execution from Python.

Thinking about this a little more, if we were to achieve that, there wouldn't necessarily be a need for a `Settings.execute` as the sequence could just be separate Python function calls. Perhaps what makes more sense in the interim is:
- Unify the current options in `Settings.run_mode` so that a user can generate a plot or volume information without using the command line, i.e. the available options would be `{'eigenvalue', 'fixed source', 'plot', 'volume', 'particle restart'}`.
- Move `<particles>`, etc. down one level in the XML (and retain backwards compatibility).

**Comment by nelsonag** [comment]
@paulromano while either the interim solution in your last post is reasonable, do we need the interim solution or is it not that much more work to just go all the way to the Python bindings?  I defer to your and @wbinventor's judgment on this as I have never gone through this route but it doesn't seem like a drastic amount of work to expose a few functions to C and then Python (or directly to Python with SWIG or something).

**Comment by paulromano** [comment]
Even if we were to go straight to the Python bindings today, the changes I outlined in my most recent comment are still needed, namely to decouple volume calculations from a normal simulation. I also feel the the XML changes proposed should be made either way as well.

I already have a branch implementing these changes and will PR it along with a few other small change soon.

**Comment by paulromano** [comment]
@nelsonag you're right though, it shouldn't be too bad to expose a few functions to C and Python. For Python bindings, SWIG is probably overkill- the same could be done with [ctypes](https://docs.python.org/3.6/library/ctypes.html), [cython](http://cython.org/), or [cffi](http://cffi.readthedocs.io/en/latest/).

**Comment by wbinventor** [comment]
Yeah I agree that SWIG is definitely too much at this point since I think it would expose far too many routines to Python which would easily be mis-used, leading to segfaults (*i.e.*, the routines within OpenMC must be called in a very specific order). Although powerful, an comprehensive set of Python bindings greatly increase the degrees of freedom with which with which users will try to use the API, which requires the code to be designed with foresight as a library of interacting data structures and routines. As implemented, OpenMC's control flow is highly sequential and would require some refactoring (such as what may be required to turn the codebase into a usable library). Alternatively, as @paulromano suggested, preferential bindings for specific routines would help alleviate this issue.

I think the general plan you outlined above @paulromano make sense - define a C/C++ API to accomplish what you need for Nek/MOOSE coupling, then expose it as Python bindings if it would be useful. I do think that a series of Python calls is preferable to a `Settings.execute` list, albeit I may be hardwired to think about things that way after working with C/C++ libraries with Python bindings like OpenMOC, SciKit-learn and Theano.

**Comment by paulromano** [comment]
Closed with #790
